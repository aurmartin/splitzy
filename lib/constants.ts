export const primaryHue = 200;

export const Colors = {
  primaryDark: `hsl(${primaryHue} 80% 40%)`,
  primary: `hsl(${primaryHue} 40% 50%)`,
  primaryLight: `hsl(${primaryHue} 40% 80%)`,

  secondary: `hsl(0 0% 70%)`,
  secondaryLight: `hsl(0 0% 80%)`,

  red: `hsl(0 50% 40%)`,

  green: `hsl(120 50% 40%)`,

  background: `hsl(0 0% 94%)`,
};
