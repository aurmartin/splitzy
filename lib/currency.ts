import { type Dinero, type Currency } from "dinero.js";

const MINOR_UNIT_FACTOR: Partial<Record<Currency, number>> = {
  EUR: 100,
  USD: 100,
};

const getAmountColor = (amount: Dinero): string => {
  if (amount.getAmount() > 0) return "hsl(120 70% 30%)";
  if (amount.getAmount() < 0) return "hsl(0 70% 30%)";
  return "black";
};

const getAbsoluteAmount = (amount: Dinero): Dinero => {
  return amount.getAmount() < 0 ? amount.multiply(-1) : amount;
};

const getBalanceMessage = (amount: Dinero): string => {
  if (amount.getAmount() > 0) return "On vous doit";
  if (amount.getAmount() < 0) return "Vous devez";
  return "Vous êtes à jour";
};

const getMinorUnitAmount = (amount: number, currency: Currency): number => {
  const factor = MINOR_UNIT_FACTOR[currency];
  if (!factor) {
    throw new Error(`Unsupported currency: ${currency}`);
  }
  return amount * factor;
};

const getMajorUnitAmount = (amount: number, currency: Currency): number => {
  const factor = MINOR_UNIT_FACTOR[currency];
  if (!factor) {
    throw new Error(`Unsupported currency: ${currency}`);
  }
  return amount / factor;
};

export {
  getAmountColor,
  getAbsoluteAmount,
  getBalanceMessage,
  getMinorUnitAmount,
  getMajorUnitAmount,
};
