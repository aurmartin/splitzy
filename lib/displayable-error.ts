export class DisplayableError extends Error {
  title: string;
  message: string;

  constructor(title: string, message: string) {
    super(`${title}: ${message}`);
    this.title = title;
    this.message = message;
  }
}
