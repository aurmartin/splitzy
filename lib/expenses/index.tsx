export * from "./types";
export * from "./hooks";

export { ExpensesRepository } from "./repository";
export { ExpenseService } from "./service";
