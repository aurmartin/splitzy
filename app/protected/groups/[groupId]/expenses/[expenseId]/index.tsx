import Label from "@/components/label";
import { ListGroup } from "@/components/list-group";
import { Screen } from "@/components/screen";
import { useSnackBar } from "@/components/snack-bar";
import { useSystem } from "@/components/system-provider";
import { Text } from "@/components/text";
import { TopBar } from "@/components/top-bar";
import { TopBarAction, TopBarDeleteAction } from "@/components/top-bar-action";
import {
  ExpenseService,
  useAmounts,
  useExpense,
  type Expense,
} from "@/lib/expenses";
import { getLocale } from "@/lib/locale";
import { router, useLocalSearchParams } from "expo-router";
import { ScrollView, View } from "react-native";

interface ShowExpenseScreenProps {
  groupId: string;
  expense: Expense;
}

function ShowExpenseScreen({ groupId, expense }: ShowExpenseScreenProps) {
  const snackBar = useSnackBar();
  const system = useSystem();
  const amounts = useAmounts(expense);

  const handleDelete = async () => {
    try {
      await ExpenseService.deleteExpense(system, expense.id);

      router.back();

      snackBar.show("Dépense supprimée", "success");
    } catch (error) {
      console.error("Delete expense error", error);

      snackBar.show("Une erreur est survenue", "error");
    }
  };

  const handleEdit = () => {
    router.replace(`/protected/groups/${groupId}/expenses/${expense.id}/edit`);
  };

  return (
    <Screen>
      <TopBar
        rightActions={
          <>
            <TopBarDeleteAction onPress={handleDelete} />
            <TopBarAction iconName="pencil" onPress={handleEdit} />
          </>
        }
      />

      <Text type="headlineMedium" style={{ marginBottom: 16 }}>
        {expense.title}{" "}
        {expense.splitExpense.total.setLocale(getLocale()).toFormat()}
      </Text>

      <Text style={{ marginBottom: 16 }}>
        Payé par {expense.payerName} le{" "}
        {new Date(expense.createdAt).toLocaleDateString()}
      </Text>

      <ScrollView
        showsVerticalScrollIndicator={false}
        alwaysBounceVertical={false}
      >
        <Label>Répartition</Label>
        <ListGroup style={{ marginBottom: 16 }}>
          {Object.entries(amounts).map(([name, amount]) => (
            <View key={name} style={{ justifyContent: "space-between" }}>
              <Text key={name}>{name}</Text>
              <Text>{amount.setLocale(getLocale()).toFormat()}</Text>
            </View>
          ))}
        </ListGroup>

        {expense.receipt && (
          <>
            <Label>Ticket de caisse</Label>
            <ListGroup itemHeight={70}>
              {expense.receipt.items.map((item) => (
                <View key={item.description}>
                  <View style={{ gap: 4, flex: 1, marginRight: 16 }}>
                    <Text
                      style={{ flexShrink: 1 }}
                      numberOfLines={1}
                      ellipsizeMode="middle"
                    >
                      {item.humanReadableDescription}
                    </Text>
                    <Text
                      type="bodyMedium"
                      style={{ color: "hsl(0, 0%, 40%)" }}
                    >
                      {item.description}
                    </Text>
                  </View>

                  <Text>{item.price.setLocale(getLocale()).toFormat()}</Text>
                </View>
              ))}
            </ListGroup>
          </>
        )}
      </ScrollView>
    </Screen>
  );
}

export default function ShowExpenseRoute() {
  const { expenseId, groupId }: { expenseId: string; groupId: string } =
    useLocalSearchParams();

  const expense = useExpense(expenseId);

  if (!expense) {
    return null;
  }

  return <ShowExpenseScreen groupId={groupId} expense={expense} />;
}
