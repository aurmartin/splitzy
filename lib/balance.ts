import dinero, { type Dinero } from "dinero.js";
import { type Expense, ExpenseService, useExpenses } from "@/lib/expenses";
import { type Group } from "@/lib/groups";

export type Balance = Record<string, Dinero>;

export type Reimbursement = {
  from: string;
  to: string;
  amount: Dinero;
};

export function computeBalance(group: Group, expenses: Expense[]): Balance {
  const balance = group.members.reduce((acc, member) => {
    acc[member] = dinero({ amount: 0, currency: group.currency });
    return acc;
  }, {} as Balance);

  expenses.forEach((expense) => {
    const payer = expense.payerName;
    const total = expense.splitExpense.total;

    balance[payer] = balance[payer].add(total);

    const amounts = ExpenseService.getAmounts(expense);

    Object.entries(amounts).forEach(([name, amount]) => {
      balance[name] = balance[name].subtract(amount);
    });
  });

  return balance;
}

export const useBalance = (group: Group): Balance => {
  const expenses = useExpenses(group.id);
  return computeBalance(group, expenses);
};

export function computeOptimalReimbursements(
  balance: Balance,
): Reimbursement[] {
  const reimbursements: Reimbursement[] = [];
  const members = Object.keys(balance);

  // Create a copy of the balance object
  const workingBalance = members.reduce((acc, member) => {
    acc[member] = balance[member];
    return acc;
  }, {} as Balance);

  // Continue while there are non-zero balances
  while (members.some((member) => !workingBalance[member].isZero())) {
    // Find the person who owes the most (most negative balance)
    const maxDebtor = members.reduce((a, b) =>
      workingBalance[a].lessThan(workingBalance[b]) ? a : b,
    );

    // Find the person who is owed the most (most positive balance)
    const maxCreditor = members.reduce((a, b) =>
      workingBalance[a].greaterThan(workingBalance[b]) ? a : b,
    );

    // If both balances are effectively zero (less than 1 cent), break
    if (
      workingBalance[maxDebtor].isZero() &&
      workingBalance[maxCreditor].isZero()
    ) {
      break;
    }

    // Calculate the reimbursement amount (minimum of absolute values)
    const debtorAmount = workingBalance[maxDebtor].multiply(-1); // Make positive
    const creditorAmount = workingBalance[maxCreditor];
    const amount = debtorAmount.lessThan(creditorAmount)
      ? debtorAmount
      : creditorAmount;

    // Update the balances
    workingBalance[maxDebtor] = workingBalance[maxDebtor].add(amount);
    workingBalance[maxCreditor] = workingBalance[maxCreditor].subtract(amount);

    // Record the reimbursement
    reimbursements.push({
      from: maxDebtor,
      to: maxCreditor,
      amount,
    });
  }

  return reimbursements;
}
