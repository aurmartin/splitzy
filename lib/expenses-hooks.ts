import { useSystem } from "@/components/system-provider";
import { Expense } from "@/lib/expenses";
import { ExpensesRepository } from "@/lib/expenses-repository";
import { useEffect, useState, useMemo } from "react";

function useExpenses(groupId: string): Expense[] {
  const system = useSystem();
  const initialExpenses = useMemo(
    () => ExpensesRepository.getGroupExpenses(system, groupId),
    [system, groupId],
  );
  const [expenses, setExpenses] = useState<Expense[]>(initialExpenses);

  useEffect(() => {
    return ExpensesRepository.watchGroupExpenses(system, groupId, setExpenses);
  }, [system, groupId]);

  return expenses;
}

function useExpense(id: string): Expense | null {
  const system = useSystem();
  const initialExpense = useMemo(
    () => ExpensesRepository.getExpense(system, id),
    [system, id],
  );
  const [expense, setExpense] = useState<Expense | null>(initialExpense);

  useEffect(() => {
    return ExpensesRepository.watchExpense(system, id, setExpense);
  }, [system, id]);

  return expense;
}

export { useExpense, useExpenses };
