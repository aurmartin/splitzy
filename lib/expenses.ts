import { convertSplit } from "@/components/split-input";
import { useSystem } from "@/components/system-provider";
import { ValidationError } from "@/lib/validation-error";
import { ExpensesRepository } from "@/lib/expenses-repository";
import { System } from "@/lib/system";
import { generateId } from "@/lib/utils";
import { Currency, Dinero } from "dinero.js";
import { useCallback } from "react";

interface BaseSplit {
  type: "percentage" | "amount" | "equal" | "receipt";
  total: Dinero;
  members: string[];
}

interface PercentageSplit extends BaseSplit {
  type: "percentage";
  ratios: { [member: string]: number };
}

interface EqualSplit extends BaseSplit {
  type: "equal";
}

interface AmountSplit extends BaseSplit {
  type: "amount";
  amounts: { [member: string]: Dinero };
}

interface ReceiptSplit extends BaseSplit {
  type: "receipt";
  receipt: Receipt;
}

type Split = PercentageSplit | EqualSplit | AmountSplit | ReceiptSplit;

interface Receipt {
  date: Date;
  total: Dinero;
  title: string;
  currency: Currency;

  items: {
    description: string;
    humanReadableDescription: string;
    price: Dinero;
    paid_for: string[];
  }[];
}

interface Expense {
  id: string;
  groupId: string;
  title: string;
  payerName: string;
  splitExpense: Split;

  createdAt: Date;
  updatedAt: Date;
  deletedAt: Date | undefined;

  receipt: Receipt | null;
}

interface ExpenseCreateParams {
  groupId: Expense["groupId"];
  title: Expense["title"];
  payerName: Expense["payerName"];
  splitExpense: Expense["splitExpense"];
  receipt: Expense["receipt"];
}

interface ExpenseUpdateParams {
  title: Expense["title"];
  payerName: Expense["payerName"];
  splitExpense: Expense["splitExpense"];
  receipt: Expense["receipt"];
}

const useAddExpense = () => {
  const system = useSystem();

  return useCallback(
    (params: ExpenseCreateParams) => addExpense(system, params),
    [system],
  );
};

const useEditExpense = () => {
  const system = useSystem();
  return useCallback(
    (id: string, params: ExpenseUpdateParams) =>
      editExpense(system, id, params),
    [system],
  );
};

const useDelExpense = () => {
  const system = useSystem();
  return useCallback(
    (id: string) => ExpensesRepository.deleteExpense(system, id),
    [system],
  );
};

const addExpense = async (system: System, params: ExpenseCreateParams) => {
  const validParams = validateCreateExpenseParams(params);
  const expense = createExpense(validParams);
  await ExpensesRepository.insertExpense(system, expense);
};

const editExpense = async (
  system: System,
  id: string,
  params: ExpenseUpdateParams,
) => {
  const expense = ExpensesRepository.getExpense(system, id);
  if (expense === null) {
    throw new Error("Expense not found");
  }

  const validParams = validateUpdateExpenseParams(params);
  const updatedExpense = modifyExpense(expense, validParams);
  await ExpensesRepository.updateExpense(system, updatedExpense);
};

const createExpense = (validParams: ExpenseCreateParams): Expense => {
  return {
    ...validParams,
    id: generateId(),
    createdAt: new Date(),
    updatedAt: new Date(),
    deletedAt: undefined,
  };
};

const modifyExpense = (
  expense: Expense,
  params: ExpenseUpdateParams,
): Expense => {
  return {
    ...expense,
    ...params,
    updatedAt: new Date(),
  };
};

const validateCreateExpenseParams = (params: ExpenseCreateParams) => {
  const errors: Record<string, string> = {};

  if (params.title.length === 0) {
    errors.title = "Veuillez entrer un titre";
  }

  if (params.payerName.length === 0) {
    errors.payerName = "Veuillez sélectionner un membre";
  }

  if (Object.keys(errors).length > 0) {
    throw new ValidationError(errors);
  }

  return params;
};

const validateUpdateExpenseParams = (params: ExpenseUpdateParams) => {
  const errors: Record<string, string> = {};

  if (params.title.length === 0) {
    errors.title = "Veuillez entrer un titre";
  }

  if (params.payerName.length === 0) {
    errors.payerName = "Veuillez sélectionner un membre";
  }

  return params;
};

const getAmountSplit = (expense: Expense): AmountSplit => {
  return convertSplit(expense.splitExpense);
};

const useExpense = ExpensesRepository.useExpense;
const useExpenses = ExpensesRepository.useExpenses;

export type {
  AmountSplit,
  EqualSplit,
  Expense,
  PercentageSplit,
  Receipt,
  ReceiptSplit,
  Split,
};

export {
  addExpense,
  getAmountSplit,
  useAddExpense,
  useDelExpense,
  useEditExpense,
  useExpense,
  useExpenses,
};
