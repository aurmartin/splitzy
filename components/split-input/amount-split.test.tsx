import {
  AmountSplitInput,
  createAmountSplit,
} from "@/components/split-input/amount-split";
import { type AmountSplit } from "@/lib/expenses";
import {
  type RenderResult,
  fireEvent,
  render,
  screen,
} from "@testing-library/react-native";
import dinero from "dinero.js";
import * as React from "react";

describe("AmountSplit", () => {
  describe("AmountSplitInput", () => {
    const testSplit = {
      type: "amount" as const,
      amounts: { alice: dinero({ amount: 1000, currency: "USD" }) },
      total: dinero({ amount: 1000, currency: "USD" }),
      members: ["alice", "bob", "charlie"],
    } as AmountSplit;

    const onChange = jest.fn();

    let tree: RenderResult;

    beforeEach(() => {
      tree = render(
        <AmountSplitInput
          testID="amount-split-input"
          value={testSplit}
          onChange={onChange}
        />,
      );
    });

    it("renders correctly", () => {
      expect(screen).toMatchSnapshot();
    });

    it("should reset amounts when the reset button is pressed", async () => {
      fireEvent.press(screen.getByTestId("reset-amount-alice"));
      expect(onChange).toHaveBeenCalledWith({ ...testSplit, amounts: {} });
    });

    it("should only display reset buttons for amounts that have been set", () => {
      expect(screen.queryByTestId("reset-amount-bob")).toBeNull();
    });

    it("should recompute remaining amounts when total changes", () => {
      const updatedSplit = {
        ...testSplit,
        total: dinero({ amount: 2000, currency: "USD" }),
      };

      expect(screen.getByTestId("amount-input-bob").props.value).toEqual("0");

      tree.rerender(
        <AmountSplitInput
          testID="amount-split-input"
          value={updatedSplit}
          onChange={onChange}
        />,
      );

      expect(screen.getByTestId("amount-input-bob").props.value).toEqual("5");
    });
  });

  describe("createAmountSplit", () => {
    it("creates equal splits correctly", () => {
      const total = dinero({ amount: 2000, currency: "USD" });
      const members = ["user1", "user2"];

      const result = createAmountSplit(total, members);

      expect(result.type).toBe("amount");
      expect(result.amounts).toEqual({});
    });

    it("handles no members", () => {
      const total = dinero({ amount: 1000, currency: "USD" });
      const members = [] as string[];

      const result = createAmountSplit(total, members);

      expect(result.type).toBe("amount");
      expect(result.amounts).toEqual({});
    });
  });
});
