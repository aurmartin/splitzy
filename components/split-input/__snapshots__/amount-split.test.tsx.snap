// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountSplit AmountSplitInput renders correctly 1`] = `
<View
  style={
    [
      {
        "gap": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderRadius": 8,
          "flexDirection": "row",
          "height": 55,
          "paddingHorizontal": 16,
        },
        {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderBottomLeftRadius": 0,
          "borderBottomRightRadius": 0,
          "borderTopLeftRadius": 8,
          "borderTopRightRadius": 8,
          "flexDirection": "row",
          "height": 55,
          "padding": 0,
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
          },
        ]
      }
    >
      alice
    </Text>
    <View
      style={
        {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <TextInput
        accessibilityLabel="alice"
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="0.00"
        placeholderTextColor="#888"
        style={
          {
            "flex": 1,
            "fontSize": 16,
            "paddingBottom": 0,
            "paddingTop": 0,
            "textAlign": "right",
          }
        }
        testID="amount-input-alice"
        value="10"
      />
    </View>
    <View
      style={
        {
          "marginLeft": 2,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="reset-amount-alice"
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "black",
                "fontSize": 18,
              },
              undefined,
              {
                "fontFamily": "material",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          Óóç
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderRadius": 8,
          "flexDirection": "row",
          "height": 55,
          "paddingHorizontal": 16,
        },
        {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderBottomLeftRadius": 0,
          "borderBottomRightRadius": 0,
          "borderTopLeftRadius": 0,
          "borderTopRightRadius": 0,
          "flexDirection": "row",
          "height": 55,
          "padding": 0,
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
          },
        ]
      }
    >
      bob
    </Text>
    <View
      style={
        {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <TextInput
        accessibilityLabel="bob"
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="0.00"
        placeholderTextColor="#888"
        style={
          {
            "flex": 1,
            "fontSize": 16,
            "paddingBottom": 0,
            "paddingTop": 0,
            "textAlign": "right",
          }
        }
        testID="amount-input-bob"
        value="0"
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderRadius": 8,
          "flexDirection": "row",
          "height": 55,
          "paddingHorizontal": 16,
        },
        {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderBottomLeftRadius": 8,
          "borderBottomRightRadius": 8,
          "borderTopLeftRadius": 0,
          "borderTopRightRadius": 0,
          "flexDirection": "row",
          "height": 55,
          "padding": 0,
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
          },
        ]
      }
    >
      charlie
    </Text>
    <View
      style={
        {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <TextInput
        accessibilityLabel="charlie"
        keyboardType="numeric"
        onChangeText={[Function]}
        placeholder="0.00"
        placeholderTextColor="#888"
        style={
          {
            "flex": 1,
            "fontSize": 16,
            "paddingBottom": 0,
            "paddingTop": 0,
            "textAlign": "right",
          }
        }
        testID="amount-input-charlie"
        value="0"
      />
    </View>
  </View>
</View>
`;
